src/rgw/rgw_rest_pubsub.cc

```diff
@@ -282,17 +291,23 @@ class RGWPSCreateTopicOp : public RGWOp {
           << "', with error:" << ret << dendl;
       return ret;
     } else {
-      existing = std::move(result);
+      topic = std::move(result);
     }
-    return RGWOp::init_processing(y);
+    return 0;
   }
 
   int verify_permission(optional_yield y) override {
-    if (!existing) {
+    if (s->auth.identity->get_account()) {
+      // account users don't consult the existing owner/policy
+      if (!verify_user_permission(this, s, topic_arn,
```

Comment:
i have couple of questions for this commit where we are adding the call `verify_user_permission`. If i understand it correctly this call is to verify that other users from same account that owns the topic have the permission (more like identity_policy)
and if this same as `identity_policy`, then we are already doing the identity check inside the `verify_topic_permission` where we check all 3, resource, identity and session. so then again calling `verify_user_permission` is required???

Second question if this `verify_user_permission` is required, then why only check this for CreateTopic, DeleteTopic and ListTopic and not check this for GetTopic,GetTopicAttr & SetTopicAttr ??

Reply:
you're right that `verify_user_permission()` is evaluating identity policies the same way as `verify_topic_permission()`

my intent was to only add `verify_user_permission()` calls in paths that weren't already calling `verify_topic_permission()` - these cases only show up CreateTopic, DeleteTopic and ListTopic

Reply:
> you're right that `verify_user_permission()` is evaluating identity policies the same way as `verify_topic_permission()`
> 
> my intent was to only add `verify_user_permission()` calls in paths that weren't already calling `verify_topic_permission()` - these cases only show up in CreateTopic, DeleteTopic and ListTopic

but from the code we see that all the topic related s3 calls do call `verify_topic_permission()` , so then we do not require explicit call to `verify_user_permission` ?

Reply:
we can only call `verify_topic_permission()` if we have a topic and know its owner/policy. both CreateTopic and DeleteTopic have to handle the case where no topic exists, so we use `verify_user_permission()` to cover those

for ListTopics, account users call `verify_user_permission()` once up front, while non-account users have to call  `verify_topic_permission()` on each topic to see which they have access to